// @fhevm-sdk/core/src/config.ts
import type { NetworkPreset, NetworkPresets } from './types';

export const NETWORK_PRESETS: Record<NetworkPreset, NetworkPresets | null> = {
  localhost: {
    aclAddress: '0x0000000000000000000000000000000000000000',
    kmsVerifierAddress: '0x0000000000000000000000000000000000000000',
    inputVerifierAddress: '0x0000000000000000000000000000000000000000',
    verifyingContractAddress: '0x0000000000000000000000000000000000000000',
    gatewayUrl: 'http://localhost:8545',
    gatewayChainId: 31337,
  },
  sepolia: {
    // Real Sepolia FHEVM testnet addresses from Zama
    aclAddress: '0x687820221192C5B662b25367F70076A37bc79b6c',
    kmsVerifierAddress: '0x1364cBBf2cDF5032C47d8226a6f6FBD2AFCDacAC',
    inputVerifierAddress: '0xbc91f3daD1A5F19F8390c400196e58073B6a0BC4',
    verifyingContractAddress: '0x7048C39f048125eDa9d678AEbaDfB22F7900a29F',
    gatewayUrl: 'https://relayer.testnet.zama.cloud',
    gatewayChainId: 55815,
  },
  mainnet: {
    aclAddress: '0x0000000000000000000000000000000000000000',
    kmsVerifierAddress: '0x0000000000000000000000000000000000000000',
    inputVerifierAddress: '0x0000000000000000000000000000000000000000',
    verifyingContractAddress: '0x0000000000000000000000000000000000000000',
    gatewayUrl: 'https://gateway.mainnet.zama.ai',
    gatewayChainId: 1,
  },
  custom: null, // User must provide all config values
};

export function resolveNetworkPresets(network: NetworkPreset): NetworkPresets {
  const preset = NETWORK_PRESETS[network];
  if (!preset) {
    throw new Error(
      `Network preset '${network}' requires custom configuration. ` +
        'Please provide aclAddress, kmsVerifierAddress, and gatewayUrl.'
    );
  }
  return preset;
}
